TERRAFORM.md

```diff
@@ -0,0 +1,103 @@
+
+## Code Organization
+
+The high level code organization should look like this
+
+```text
+|- modules/                   # terraform reusable modules for the whole repository
+|- ORGANIZATIONS_OR_ACCOUNT   # e.g. AWS Account name
+|           |- modules        # terraform reusable modules for this organization/account
+|           |- shared         # shared resources
+|           |- ENVIRONMENT    # resource pro environment
+```
+
+Then it should mirror the hierarchical structure of the infrastructure. Note that each terraform module (project) should be kept small.
+
+### Reusable Module Usage
+
+Use modules to encapsulate and reuse configurations. This promotes DRY (Don't Repeat Yourself) principles and makes your configurations more manageable.
+
+- Create modules for common resources.
+- Use versioned modules from the Terraform Registry when possible.
+- Use modules to share global variales between terraform modules
+
+## State Management
+
+### Remote State
+
+We use remote state management using AWS S3 backend
+
+- Encrypt the state file at rest (see [Terraform Documentation](https://developer.hashicorp.com/terraform/language/settings/backends/s3#encrypt)).
+- Put the state backend definition in `terraform.tf` file
+
+#### State key
+
+We use the terraform module definition file path as state key in S3 prefixed with a version string `v1`; e.g. `v1/infra-terraform-bgdi/systems-prod/map.terraform.tf`
+
+The version prefix is to avoid any risk of key collision in the future in case of code structure re-organization.
+
+*NOTE: We used to use an UUID as state keys, which has the advantage that we did not have to change the state when reorganizing resources. However it has the downside that UUID key were very hard to code review and error prone due to copy paste mistake not being easily identifiable.*
+
+### State Locking
+
+Enable state locking to prevent concurrent operations from corrupting your state.
+
+- Use AWS DynamoDB to manage state locks.
+
+### State migration
+
+To migrate the state we use the `terraform init -migrate-state` command. After migrating the state we need to manually clean up the old state by removing the S3 file and deleting the dynamoDB entry
+
+Here is an example of migrate state procedure
+
+1. Run the `terraform init` BEFORE changing the state key on the terraform backend file :warning:
+2. Update the the state key in the terraform backend file
+3. Run `terraform init -migrate-state`
+4. Clean the old state in S3 and DynamoDB
+
+```bash
+OLD_KEY=<old-state-key>
+
+aws --profile AWS_ACCOUNT_PROFILE s3 rm s3://STATE_BUCKET/${OLD_KEY}
+
+aws --profile AWS_ACCOUNT_PROFILE dynamodb delete-item --table-name TABLE_NAME --key '{"LockID": {"S": "STATE_BUCKET/'${OLD_KEY}'-md5"}}'
+```
+
+### Module Guidelines
+
+- :warning: ALWAYS change resources using terraform (no manual changes via UI, e.g. AWS web console) :warning:
+- :warning: NEVER EVER apply changes before opening a Pull Request :warning:
```

Comment:
Depending how you read this, I don't think that's how we currently do it, neither do I think that's really tractable. At least in dev. The workflow I have been using is:
1. prepare a first draft of my change, try to check it does the right thing with "terraform plan"
2. open a PR
3. maybe get someone to look at the PR, maybe have it approved
4.  "terraform apply" the change
5. fix what doesn't work, go back to 4
6. once it does the right thing, push my chanages to the PR
7. get the PR approved
8. merge the PR
9. maybe verify "terraform apply" in master doesn't show any change

In the cycle between 4 and 5, the state of the PR and the state of what I "terraform apply" with be inconsistent.

I think the important part is to make sure that anybody who runs into an inconsistent state needs to be able to easily track down what's going on. Hence, it is important to have a PR open that people can find even if its state is not 100% consistent all the time. And to not leave changes that have been terraform-applied but not github-merged hanging for too long. I think this change explains the last part but maybe it can do a better job at explaining the first part (the "why").

Also, maybe if we can have a bot that announce Terraform PRs on Slack that would be handy?

Also also, this whole section sounds more like "development guidelines" than "module guidelines".

Reply:
Here we could solve all this by using Atlantis https://www.runatlantis.io/ but it was decided that for the moment we don't want another tool to manage

Reply:
Sure. But for now can we explain here why it's important to open a PR showing the intent of the change before applying the change? I think it's to allow other to easily track down what's happening in case of inconsistency/conflict but there may be other reasons?